### PR TITLE
Check for GPS device in the phone and show dialog if not present.

### LINKED ISSUE #876: Detect absence of GPS unit
Some phones doesn't have a GPS unit

The older version of the app has some logic to detect that there is no GPS unit.   This can be determined by asking the location service which providers are available.

https://github.com/Greenstand/treetracker-android/tree/release

### PR DESCRIPTION
**Pull request by EngrTaofeek**
fix #876 Check for GPS device in the phone and show dialog if not present. Clicking the button in the dialog closes the app.
Screenshot_20220307-191728

### CODE REVIEW

File: app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt [hunk 1]
```diff
@@ -26,11 +37,36 @@ class TreeTrackerActivity : ComponentActivity() {
...
+                }
             }
         }
     }
+
+    fun hasGPSDevice(context: Context): Boolean {
+        val mgr = context.getSystemService(LOCATION_SERVICE) as LocationManager
+        val providers = mgr.allProviders
+        return providers.contains(LocationManager.GPS_PROVIDER)
+    }
```
Comment (Elforama) [inline app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt:61]: Let's try to keep the activity as clean as possible. It's easy for activities to turn into super large classes as time goes on. We can move this logic into one of the Location classes we have.

File: app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt [hunk 2]
```diff
@@ -26,11 +37,36 @@ class TreeTrackerActivity : ComponentActivity() {
...
+}
+
+@Composable
+fun NoGPSDeviceDialog(onPositiveClick: () -> Unit){
+    CustomDialog(
+        dialogIcon = painterResource(id = R.drawable.error_outline),
+        title = stringResource(R.string.no_gps_device_header),
+        textContent = stringResource(R.string.no_gps_device_content),
+        onPositiveClick = onPositiveClick
+    )
```
Comment (Elforama) [inline app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt:71]: Please move this to one of the dialog files, or create a new one.

File: app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt [hunk 3]
```diff
@@ -26,11 +37,36 @@ class TreeTrackerActivity : ComponentActivity() {
         } else {
             languageSwitcher.applyCurrentLanguage(this)
         }
-
-        setContent {
-            CustomTheme {
-                Root(viewModelFactory)
+        if (hasGPSDevice(this)) {
```
Comment (Elforama) [inline app/src/main/java/org/greenstand/android/TreeTracker/activities/TreeTrackerActivity.kt:40]: It looks like the main difference here is Root vs NoGPSDiviceDialog. We can move the if statement inside the custom theme.

File: app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -104,4 +104,6 @@
     <string name="selfie_instruction">to take a selfie to continue your registration.</string>
     <string name="recapture_tutorial">Click on the button to try capturing the tree again.</string>
     <string name="capture_tutorial">Click on the button to save tree.</string>
+    <string name="no_gps_device_header">NO GPS DEVICE FOUND</string>
+    <string name="no_gps_device_content">This app is not supported on this device because no GPS receiver was found. Kindly try another phone that has GPS device</string>
```
Comment (Elforama) [inline app/src/main/res/values/strings.xml:108]: I think we can reword this to say something more like 
"This device does not support GPS. Please use a device that has GPS to continue."

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -104,4 +104,6 @@
     <string name="selfie_instruction">to take a selfie to continue your registration.</string>
     <string name="recapture_tutorial">Click on the button to try capturing the tree again.</string>
     <string name="capture_tutorial">Click on the button to save tree.</string>
+    <string name="no_gps_device_header">NO GPS DEVICE FOUND</string>
```
Comment (Elforama) [inline app/src/main/res/values/strings.xml:107]: This can probably say  "No GPS found on device"

**Review by Elforama — CHANGES_REQUESTED** [review]
Looks good, but needs some cleanup, and possible rewording of the dialog.

File: app/src/main/java/org/greenstand/android/TreeTracker/models/location/Locations.kt [hunk 6]
```diff
@@ -27,3 +51,9 @@ data class LocationData(
     val convergenceStatus: ConvergenceStatus?,
     val capturedAt: String
 )
+
+fun hasGPSDevice(context: Context): Boolean {
```
Comment (Elforama) [inline app/src/main/java/org/greenstand/android/TreeTracker/models/location/Locations.kt:55]: Please move this into a util class to either use statically, or to inject

**Review by Elforama — CHANGES_REQUESTED** [review]
Looks good. Please just address the comment about the GPS function

File: app/src/main/res/values/strings.xml (hunk 4 above)
Comment (ZavenArra) [reply app/src/main/res/values/strings.xml]: We also need a workflow for making sure we have a swahili translation for any new text added.  @sebastiangaertner can provide this.  Please file a ticket for him.

File: app/src/main/res/values/strings.xml (hunk 5 above)
Comment (ZavenArra) [reply app/src/main/res/values/strings.xml]: same here, Swahili needed @sebastiangaertner

File: app/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -104,4 +104,6 @@
     <string name="selfie_instruction">to take a selfie to continue your registration.</string>
     <string name="recapture_tutorial">Click on the button to try capturing the tree again.</string>
     <string name="capture_tutorial">Click on the button to save tree.</string>
+    <string name="no_gps_device_header">No GPS found on device</string>
+    <string name="no_gps_device_content">This device does not support GPS. Please use a device that has GPS to continue.</string>
```
Comment (ZavenArra) [inline app/src/main/res/values/strings.xml:108]: Always need a swahili translation of any new text added.

**Comment by EngrTaofeek** [comment]
This PR has been moved to #967
